hands-on-exercise/1-native/5-add-query.md

```diff
@@ -0,0 +1,310 @@
+// method.
+message QueryGetGameResponse {
+  // Game defines the game at the requested index.
+  StoredGame Game = 1;
+}
+```
+
+Note that the response reuses `StoredGame` for convenience and to keep this exercise small. In fact you are free to define a new type to return. For instance this new type could be like `StoredGame` minus the `Index` field as the index is implicit over the request's lifecycle.
+
+### Compile Protobuf
+
+Since you have defined a new query type and an associated `service`, you should recompile the lot:
+
+```sh
+$ make proto-gen
+```
+
+In the new `query.pg.go` you can find `type QueryGetGameRequest struct` and `type QueryServer interface`. Now you can use them closer to the keeper.
+
+### New query server
+
+Create a new file `keeper/query_server.go` and take inspiration from `minimal-module-example`. It needs to:
+
+* Implement the `GetGame` function.
+* Return the game if it is found.
+* Return no games and no errors if it is not found.
+* Return an error otherwise.
+
+```go [keeper/query_server.go]
+package keeper
+
+import (
+    "context"
+    "errors"
+
+    "cosmossdk.io/collections"
+    "google.golang.org/grpc/codes"
+    "google.golang.org/grpc/status"
+
+    "github.com/alice/checkers"
+)
+
+var _ checkers.QueryServer = queryServer{}
+
+// NewQueryServerImpl returns an implementation of the module QueryServer.
+func NewQueryServerImpl(k Keeper) checkers.QueryServer {
+    return queryServer{k}
+}
+
+type queryServer struct {
+    k Keeper
+}
+
+// GetGame defines the handler for the Query/GetGame RPC method.
+func (qs queryServer) GetGame(ctx context.Context, req *checkers.QueryGetGameRequest) (*checkers.QueryGetGameResponse, error) {
+    game, err := qs.k.StoredGameList.Get(ctx, req.Index)
+    if err == nil {
+        return &checkers.QueryGetGameResponse{Game: &game}, nil
+    }
+    if errors.Is(err, collections.ErrNotFound) {
+        return &checkers.QueryGetGameResponse{Game: nil}, nil
+    }
+
+    return nil, status.Error(codes.Internal, err.Error())
+}
+```
+
+Note how the response's `Game` is a pointer, this lets you define it as `nil` in the case there is no game at the requested index.
+
+With the query server defined, you now need to register it into the module.
+
+### Register the types in the module
+
+Now that you have message types and server, you should register the service in `module/module.go`. Inspire yourself from what you find in `minimal-module-example`. The lines were previously commented out:
+
+```diff-go [module/module.go]
+    ...
+    func (AppModule) RegisterGRPCGatewayRoutes(clientCtx client.Context, mux *gwruntime.ServeMux) {
+-      // if err := checkers.RegisterQueryHandlerClient(context.Background(), mux, checkers.NewQueryClient(clientCtx)); err != nil {
+-      //     panic(err)
+-      // }
++      if err := checkers.RegisterQueryHandlerClient(context.Background(), mux, checkers.NewQueryClient(clientCtx)); err != nil {
++          panic(err)
```

Comment:
```
➜  chain-minimal git:(main) ✗ make install
--> ensure dependencies have not been modified
all modules verified
--> installing minid
# github.com/alice/checkers/module
../checkers-minimal/module/module.go:55:50: undefined: context
make: *** [Makefile:31: install] Error 1
```

One needs to add `"context"` to the import list.

Reply:
Added now.